[mlir][SPIRV] Add decorateType method for MatrixType

trdthg opened the pull request:
try fix #108161 

This PR adds a decorateType method for MatrixType, ensuring that `spirv.matrix` with offset in `spirv.struct` can be handled correctly.

kuhar reviewed (commented):
The logic looks OK, just a few minor issues

trdthg commented on mlir/lib/Dialect/SPIRV/Utils/LayoutUtils.cpp:
```diff
@@ -138,6 +140,25 @@ Type VulkanLayoutUtils::decorateType(spirv::ArrayType arrayType,
+Type VulkanLayoutUtils::decorateType(spirv::MatrixType matrixType,
+                                     VulkanLayoutUtils::Size &size,
+                                     VulkanLayoutUtils::Size &alignment) {
+  const auto numColumns = matrixType.getNumColumns();
+  const auto columnType = matrixType.getColumnType();
+  const auto numElements = matrixType.getNumElements();
+  auto elementType = matrixType.getElementType();
```
`getElementType()` will return `Type`
you mean `Type elementType = matrixType.getElementType();` would be better?
I think it's not bad, but the code in other similar places also uses `auto`. Should I replace them all?

trdthg replied:
Kindly ping for this question @kuhar

kuhar replied:
> you mean `Type elementType = matrixType.getElementType();` would be better?
Yes.

> I think it's not bad, but the code in other similar places also uses `auto`. Should I replace them all?
I'd rather have larger cleanups in a separate PR. It's fine if things are not exactly uniform, but we should converge towards code that follows the mlir coding standards.

trdthg replied:
> Yes. 

then it's done

> cleanups in a separate PR

I'll handle them later

trdthg replied:
Please check https://github.com/llvm/llvm-project/pull/113877

trdthg committed:
[mlir][SPIRV] Add decorateType method for MatrixType

This PR adds a decorateType method for MatrixType,
ensuring that `spirv.matrix` with offset in `spirv.struct`
can be handled correctly.

Signed-off-by: MingZhu Yan <yanmingzhu@iscas.ac.cn>

kuhar commented on mlir/lib/Dialect/SPIRV/Utils/LayoutUtils.cpp:
```diff
@@ -138,6 +140,25 @@ Type VulkanLayoutUtils::decorateType(spirv::ArrayType arrayType,
+Type VulkanLayoutUtils::decorateType(spirv::MatrixType matrixType,
+                                     VulkanLayoutUtils::Size &size,
+                                     VulkanLayoutUtils::Size &alignment) {
+  const auto numColumns = matrixType.getNumColumns();
+  const auto columnType = matrixType.getColumnType();
+  const auto numElements = matrixType.getNumElements();
```
use the actual types instead of `auto` since the type is not obvious based on the RHS only

trdthg replied:
done

kuhar reviewed (approved):
LGTM % nit

kuhar commented:
@trdthg seems like the code doesn't build

trdthg commented:
fixed
